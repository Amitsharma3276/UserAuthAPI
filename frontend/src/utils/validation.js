export const validate = (values) => {
  const digits_only = (string) =>
    [...string].every((ch) => "0123456789".includes(ch));

  const alphabet = (string) =>
    [...string].every((ch) =>
      "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ".includes(ch)
    );

  let retError = "";

  //NAME VALIDATIONS
  if (values.name === null || values.name.length === 0) {
    retError = retError.concat("Name is required \n");
  }

  if (!alphabet(values.name)) {
    retError = retError.concat("Invalid Name \n");
  }

  //PASSWORD VALIDATIONS
  if (values.password === "" || values.password === null) {
  } else if (
    values.password != null &&
    (values.password.length > 15 || values.password.length < 8)
  ) {
    retError = retError.concat("Weak password  \n");
  }

  //PERMANENT ADDRESS VALIDATION
  if (values.p_line_one === null || values.p_line_one.length === 0) {
    retError = retError.concat("Permanent Address Line 1 is required \n");
  }

  if (values.p_city === null || values.p_city.length === 0) {
    retError = retError.concat("Permanent Address City is required \n");
  }

  if (values.p_state === null || values.p_state.length === 0) {
    retError = retError.concat("Permanent Address State is required \n");
  }

  if (
    values.p_pincode === null ||
    values.p_pincode.length !== 6 ||
    (values.p_pincode.length != null && !digits_only(values.p_pincode))
  ) {
    retError = retError.concat(
      "Enter Valid 6 digit pincode for permanent address \n"
    );
  }

  //CURRENT ADDRESS VALIDATION(IF ENTERED)
  if (values.c_line_one !== null && values.c_line_one.length !== 0) {
    if (values.c_city === null || values.c_city.length === 0) {
      retError = retError.concat("Current Address City is required \n");
    }
    if (values.c_state === null || values.c_state.length === 0) {
      retError = retError.concat("Current Address State is required \n");
    }
    if (
      values.c_pincode === null ||
      values.c_pincode.length !== 6 ||
      (values.c_pincode.length !== 0 && !digits_only(values.c_pincode))
    ) {
      retError = retError.concat(
        "Enter Valid 6 digit pincode for current address \n"
      );
    }
  }

  //EMERGENCY NUMBER VALIDATION
  if (
    values.emergency_number === null ||
    values.emergency_number.length !== 10 ||
    (values.emergency_number != null && !digits_only(values.emergency_number))
  ) {
    retError = retError.concat("Enter valid 10 digit contact number \n");
  }

  return retError;
};

// const alphaNumeric = (string) =>
// [...string].every((ch) =>
//   "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789".includes(
//     ch
//   )
// );
