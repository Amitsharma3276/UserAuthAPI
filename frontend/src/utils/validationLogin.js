export const validationLogin = (values) => {
  let retError = "";
  console.log(values);
  if (values.username === "" || values.username === null) {
    retError = retError.concat("Enter Username  \n");
  }
  if (values.password === "" || values.password === null) {
    retError = retError.concat("Enter Password  \n");
    console.log("Hello");
  } else if (
    values.password != null &&
    (values.password.length > 15 || values.password.length < 8)
  ) {
    retError = retError.concat("Enter correct password  \n");
  }
  console.log(retError);

  return retError;
};
