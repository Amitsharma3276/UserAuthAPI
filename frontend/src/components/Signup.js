import React, { useState } from "react";
import { validate } from "../utils/validation";
import FormHeader from "../common/FormHeader";
import Address from "../common/Address";
import Input from "../common/Input";
import { signupService } from "../services/signupService";
import { Error } from "../common/Error";

const initialSignupState = {
  username: "",
  password: "",
  name: "",
  vaccinated: "true",
  emergency_number: "",
  skills: "",
  c_line_one: "",
  c_line_two: "",
  c_city: "",
  c_pincode: "",
  c_state: "",
  p_line_one: "",
  p_line_two: "",
  p_city: "",
  p_pincode: "",
  p_state: "",
  error: "",
  success: "false",
};

const Signup = () => {
  const [userInfo, setUserInfo] = useState(initialSignupState);

  const handleChange = (userAttribute) => (e) => {
    setUserInfo({ ...userInfo, error: false, [userAttribute]: e.target.value });
  };

  const signupClick = (e) => {
    e.preventDefault();
    let errorMessage = validate(userInfo);
    if (errorMessage) {
      setUserInfo({ ...userInfo, error: errorMessage, success: false });
      return;
    }

    setUserInfo({ ...userInfo, error: false });

    let status;
    signupService(userInfo)
      .then((response) => {
        //console.log(response.status);
        status = response.status;
        return response.text();
      })
      .then((data) => {
        console.log(data);

        if (status === 201) {
          setUserInfo({
            ...userInfo,
            success: true,
            username: "",
            password: "",
            name: "",
            vaccinated: "",
            emergency_number: "",
            skills: "",
            c_line_one: "",
            c_line_two: "",
            c_city: "",
            c_pincode: "",
            c_state: "",
            p_line_one: "",
            p_line_two: "",
            p_city: "",
            p_pincode: "",
            p_state: "",
            error: data,
          });
        } else {
          setUserInfo({ ...userInfo, error: data, success: false });
        }
      });
  };

  const formHeader = {
    title: <h3>Welcome to Team AAAA's webapp</h3>,
    message: <p>Please Signup with your credentials</p>,
  };

  return (
    <div className="auth-inner container-fluid">
      <form onSubmit={signupClick} id="myForm">
        <FormHeader title={formHeader.title} message={formHeader.message} />
        <br />
        {Error(userInfo)}

        <Input
          label="Name"
          onChange={handleChange("name")}
          type="text"
          className="form-control"
          placeholder="Name"
          userInfo={userInfo.name}
          isRequired="true"
        />
        <div className="row">
          <div className="col">
            <Input
              label="Username"
              onChange={handleChange("username")}
              type="text"
              className="form-control"
              placeholder="Enter username"
              value={userInfo.username}
              isRequired="false"
            />
          </div>
          <div className="col">
            <Input
              label="Password"
              onChange={handleChange("password")}
              type="password"
              className="form-control"
              placeholder="Enter password"
              value={userInfo.password}
              isRequired="false"
            />
          </div>
        </div>

        <Input
          label="Contact Number"
          onChange={handleChange("emergency_number")}
          type="text"
          className="form-control"
          placeholder="Enter your contact Number"
          value={userInfo.emergency_number}
          isRequired="true"
        />

        <hr />

        <div className="row">
          <div className="col">
            <div className="form-group">
              <p>
                <strong>Permanent Address</strong>*
              </p>
            </div>
            <Address
              kind={"p"}
              line_one={userInfo.p_line_one}
              line_two={userInfo.p_line_two}
              city={userInfo.p_city}
              pincode={userInfo.p_pincode}
              state={userInfo.p_city}
              handleChange={handleChange}
              isRequired="true"
            />
          </div>

          <div className="col">
            <div className="form-group">
              <p>
                <strong>Current Address</strong>
              </p>
            </div>
            <Address
              kind={"c"}
              line_one={userInfo.c_line_one}
              line_two={userInfo.c_line_two}
              city={userInfo.c_city}
              pincode={userInfo.c_pincode}
              state={userInfo.c_city}
              handleChange={handleChange}
              isRequired="false"
            />
          </div>
        </div>

        <hr />

        <div className="form-group">
          <label>Vaccination Status</label>
          <select
            onChange={handleChange("vaccinated")}
            type="number"
            className="form-control"
            value={userInfo.vaccinated}
          >
            <option value="true">YES</option>
            <option value="false">NO</option>
          </select>
        </div>

        <div className="form-group">
          <Input
            label="Skills"
            onChange={handleChange("skills")}
            type="text"
            className="form-control"
            placeholder="Add atleast 2 skills"
            value={userInfo.skills}
            isRequired="true"
          />
        </div>

        <button type="submit" className="btn btn-primary btn-block">
          Sign Up
        </button>
        <p className="forgot-password text-right">
          Already registered? <a href="/login">Log in</a>
        </p>
      </form>
    </div>
  );
};

export default Signup;
