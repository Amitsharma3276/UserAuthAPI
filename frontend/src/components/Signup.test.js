import React from "react";
import { shallow } from "enzyme";
import Signup from "./Signup";
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { act } from "react-dom/test-utils";
import { render } from "@testing-library/react";
import { unmountComponentAtNode } from "react-dom";
import * as signupUtil from "../utils/validation";

configure({ adapter: new Adapter() });
let container = null;
beforeEach(() => {
  //render target
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("<Signup /> with no props", () => {
  const wrapper = shallow(<Signup />);

  //SNAPSHOT
  it("should match the snapshot", () => {
    act(() => {
      render(wrapper, container);
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("Password onChangeHandler test", () => {
    act(() => {
      render(wrapper, container);
    });
    // console.log('Check#'+wrapper.find('Input[type="password"]').props('onChange'));
    wrapper.find('Input[type="password"]').simulate("change", {
      target: { value: "Dummy@123" },
    });

    expect(wrapper.find('Input[type="password"]').prop("value")).toEqual(
      "Dummy@123"
    );
  });

  //SIGNUP BUTTON
  it("render signup component", () => {
    act(() => {
      render(wrapper, container);
    });
    const button = document.querySelector("button");
    expect(button.innerHTML).toBe("Sign Up");
  });

  it("signupClick Button test with no errors", () => {
    act(() => {
      render(wrapper, container);
    });
    const clickEvent = { preventDefault: jest.fn() };
    //const spy = jest.spyOn(wrapper.instance(), "validationLogin");
    wrapper.find("button").simulate("click", clickEvent);
    expect(clickEvent.preventDefault).toBeCalledTimes(0);
  });

  it("signupClick form submit test with no errors", () => {
    act(() => {
      render(wrapper, container);
    });
    const clickEvent = { preventDefault: jest.fn() };
    //const spy = jest.spyOn(wrapper.instance(), "validationLogin");
    wrapper.find("form").simulate("submit", clickEvent);
    expect(clickEvent.preventDefault).toBeCalledTimes(1);
  });

  it("Signup Validation", () => {
    const fakeData = {
      username: "dummy",
      password: "dummy@123",
      name: "dummy",
      vaccinated: "true",
      emergency_number: "1234567890",
      skills: "c,cpp",
      c_line_one: "",
      c_line_two: "",
      c_city: "",
      c_pincode: "",
      c_state: "",
      p_line_one: "line1",
      p_line_two: "line2",
      p_city: "city",
      p_pincode: "123456",
      p_state: "state",
      error: "",
      success: "false",
    };

    const errorResult = signupUtil.validate(fakeData);
    //console.log(errorResult);
    expect(errorResult).toBe("");

    const errorMessage = "Invalid Credentials";
    const mockValidationSignUp = jest.spyOn(signupUtil, "validate");
    mockValidationSignUp.mockResolvedValue(errorMessage);
    return signupUtil.validate(fakeData).then((response) => {
      //console.log(response);
      expect(response).toBe(errorMessage);
      expect(mockValidationSignUp).toHaveBeenCalledTimes(1);
    });
  });

  // it("Login Validation with errors", () => {
  //   const fakeData = {
  //     username: "dummy",
  //     password: "",
  //     error: "",
  //     data: "",
  //     redirectToHome: false,
  //   };

  //   const errorResult1 = loginUtil.validationLogin(fakeData);
  //   console.log(errorResult1);
  //   //expect(errorResult1).toBe("Enter password \n");

  //   const erroMessage = "Enter password";
  //   const mockValidationLogin = jest.spyOn(loginUtil, "validationLogin");
  //   mockValidationLogin.mockResolvedValue(erroMessage);
  //   return loginUtil.validationLogin(fakeData).then((response) => {
  //     expect(response).toBe(erroMessage);
  //     expect(mockValidationLogin).toHaveBeenCalledTimes(2);
  //   });
  // });
});
