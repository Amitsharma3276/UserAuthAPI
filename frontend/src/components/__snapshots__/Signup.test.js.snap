// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Signup /> with no props should match the snapshot 1`] = `"<div class=\\"auth-inner container-fluid\\"><form id=\\"myForm\\"><div><center><h3>Welcome to Team AAAA&#x27;s webapp</h3><p>Please Signup with your credentials</p></center><hr/></div><br/><div class=\\"alert alert-warning display-linebreak\\" style=\\"display:none\\"><div class=\\"display-linebreak\\" style=\\"white-space:pre-line\\"></div></div><div class=\\"form-group\\"><label>Name*</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Name\\"/></div><div class=\\"row\\"><div class=\\"col\\"><div class=\\"form-group\\"><label>Username</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Enter username\\"/></div></div><div class=\\"col\\"><div class=\\"form-group\\"><label>Password</label><input type=\\"password\\" class=\\"form-control\\" placeholder=\\"Enter password\\"/></div></div></div><div class=\\"form-group\\"><label>Contact Number*</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Enter your contact Number\\"/></div><hr/><div class=\\"row\\"><div class=\\"col\\"><div class=\\"form-group\\"><p><strong>Permanent Address</strong>*</p></div><div><div class=\\"form-group\\"><label>Address Line1*</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Enter House No., Road No.\\"/></div><div class=\\"form-group\\"><label>Address Line2*</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Enter Area, Landmark\\"/></div><div class=\\"form-group\\"><label>City*</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Enter City\\"/></div><div class=\\"form-group\\"><label>State*</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Enter State\\"/></div><div class=\\"form-group\\"><label>Pincode*</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Enter your Pincode\\"/></div></div></div><div class=\\"col\\"><div class=\\"form-group\\"><p><strong>Current Address</strong></p></div><div><div class=\\"form-group\\"><label>Address Line1</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Enter House No., Road No.\\"/></div><div class=\\"form-group\\"><label>Address Line2</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Enter Area, Landmark\\"/></div><div class=\\"form-group\\"><label>City</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Enter City\\"/></div><div class=\\"form-group\\"><label>State</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Enter State\\"/></div><div class=\\"form-group\\"><label>Pincode</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Enter your Pincode\\"/></div></div></div></div><hr/><div class=\\"form-group\\"><label>Vaccination Status</label><select type=\\"number\\" class=\\"form-control\\"><option selected=\\"\\" value=\\"true\\">YES</option><option value=\\"false\\">NO</option></select></div><div class=\\"form-group\\"><div class=\\"form-group\\"><label>Skills*</label><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Add atleast 2 skills\\"/></div></div><button type=\\"submit\\" class=\\"btn btn-primary btn-block\\">Sign Up</button><p class=\\"forgot-password text-right\\">Already registered? <a href=\\"/login\\">Log in</a></p></form></div>"`;
